File: DemoApp Objective-C/CASSample/NativeAd/NativeTemplateVC.m
//
//  NativeTemplateVC.m
//  CASSample
//

#import "AppDelegate.h"
#import "NativeTemplateVC.h"

@interface NativeTemplateVC ()

@end

@implementation NativeTemplateVC

- (void)viewDidLoad {
    [super viewDidLoad];
    [self sizeSetup];
    [self loaderSetup];
    [self customizeAdViewAppearance];
    
    self.navigationItem.largeTitleDisplayMode = UINavigationItemLargeTitleDisplayModeNever;
    self.navigationItem.title = @"Native Template Ad";
}

- (void)sizeSetup {
    CASSize *size = [CASSize mediumRectangle];
    [self.adView setAdTemplateSize:size];
}

- (void)loaderSetup {
    self.nativeLoader = [[CASNativeLoader alloc] initWithCasID: [AppDelegate casId]];
    self.nativeLoader.delegate = self;
    self.nativeLoader.adChoicesPlacement = CASChoicesPlacementTopRight;
    self.nativeLoader.isStartVideoMuted = YES;
}

- (void)customizeAdViewAppearance {
    self.adView.backgroundColor = [UIColor whiteColor];
    self.adView.headlineView.textColor = [UIColor redColor];
}

- (IBAction)loadAction:(UIButton *)sender {
    [self.nativeLoader loadAd];
}

#pragma mark - CASImpressionDelegate

- (void)adDidRecordImpressionWithInfo:(CASContentInfo * _Nonnull)info {
    NSLog(@"adDidRecordImpressionWithInfo");
}

#pragma mark - CASNativeLoaderDelegate

- (void)nativeAdDidLoadContent:(CASNativeAdContent * _Nonnull)ad {
    ad.delegate = self;
    ad.impressionDelegate = self;
    ad.rootViewController = self;
    [self.adView setNativeAd:ad];
    NSLog(@"nativeAdDidLoadContent");
}

- (void)nativeAdDidFailToLoadWithError:(CASError * _Nonnull)error {
    NSLog(@"nativeAdDidFailToLoadWithError %@", error.description);
}


#pragma mark - CASNativeAdContentDelegate

- (void)nativeAd:(CASNativeAdContent * _Nonnull)ad didFailToPresentWithError:(CASError * _Nonnull)error {
    NSLog(@"nativeAd:didFailToPresentWithError %@", error.description);
}

- (void)nativeAdDidClickContent:(CASNativeAdContent * _Nonnull)ad {
    NSLog(@"nativeAdDidClickContent");
}

@end

File: DemoApp Swift/CASSample/NativeAd/NativeTemplateVC.swift
//
//  NativeTemplateVC.swift
//  CASSample
//
//  Copyright © 2025 Clever Ads Solutions. All rights reserved.
//

import UIKit
import CleverAdsSolutions

class NativeTemplateVC: UIViewController {
    
    @IBOutlet var loadButton: UIButton!
    @IBOutlet weak var adView: CASNativeView!
    
    private var nativeLoader = CASNativeLoader(casID: AppDelegate.casId)
    
    override func viewDidLoad() {
        super.viewDidLoad()
        sizeSetup()
        loaderSetup()
        customizeAdViewAppearance()
        //
        navigationItem.largeTitleDisplayMode = .never
        navigationItem.title = "Native Template Ad"
    }
    
    private func sizeSetup() {
        let size = AdSize.mediumRectangle
        adView.setAdTemplateSize(size)
    }
    
    private func loaderSetup() {
        nativeLoader.delegate = self
        nativeLoader.adChoicesPlacement = AdChoicesPlacement.topRight // by default
        nativeLoader.isStartVideoMuted = true
    }
    
    private func customizeAdViewAppearance() {
        // Default values are shown below:
        adView.backgroundColor = UIColor.white
        adView.headlineView?.textColor = UIColor.red
    }
    
    @IBAction private func loadAction(_ sender: UIButton) {
        nativeLoader.loadAd()
    }
}

extension NativeTemplateVC: CASImpressionDelegate {
    func adDidRecordImpression(info: AdContentInfo) {
        print(#function, "Ad Source: \(info.sourceName)")
    }
}

extension NativeTemplateVC: CASNativeLoaderDelegate {
    func nativeAdDidLoadContent(_ ad: NativeAdContent) {
        ad.delegate = self
        ad.impressionDelegate = self
        ad.rootViewController = self
        adView.setNativeAd(ad)
        print(#function)
    }
  
    func nativeAdDidFailToLoad(error: AdError) {
        print(#function, "Error: \(error.description)")
    }
}

extension NativeTemplateVC: CASNativeContentDelegate {
    func nativeAd(_ ad: NativeAdContent, didFailToPresentWithError error: AdError) {
        print(#function, "Error: \(error.description)")
    }
    
    func nativeAdDidClickContent(_ ad: NativeAdContent) {
        print(#function)
    }
}
